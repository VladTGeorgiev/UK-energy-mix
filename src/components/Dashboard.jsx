import React, { Component } from 'react';
import { Header, Divider, Label, Container, Dimmer, Loader, Icon, Grid } from 'semantic-ui-react';
import API from '../adapters/API.jsx';
import MixChart from './MixChart.jsx';
import FuelCard from './FuelCard.jsx';

class Dashboard extends Component {
	constructor (props) {
		super(props);
		this.state = {
			data: {},
			width: window.innerWidth
		};
	};

	componentDidMount () {
		window.addEventListener('resize', this.handleWindowSizeChange);
		API.fetchData().then((data) => {
			this.setState({
				data: data
			});
		});
	};

	componentWillUnmount () {
		window.removeEventListener('resize', this.handleWindowSizeChange);
	};

	handleWindowSizeChange = () => {
		this.setState({ width: window.innerWidth });
	};

	render () {
		const output = this.state.data.data === undefined ? null : this.state.data.data;
		const dashboardLayout = this.state.width < 1000 ? 1 : 2;

		if (this.state.data.error) {
			return (
				<Container textAlign="center">
					<Divider hidden />
					<Header as="h1">UK Energy Mix</Header>
					<Divider hidden />
					<Label size="big" color="red">
						{this.state.data.error.code}
					</Label>
					<Divider hidden />
					<Label>{this.state.data.error.message}</Label>
				</Container>
			);
		} else {
			return (
				<Container textAlign="center">
					<Divider hidden />
					<Header as="h1">UK Energy Mix</Header>
					<Divider />
					{output === null ? (
						<Container>
							<Divider hidden />
							<Dimmer active inverted>
								<Loader size="large">Loading...</Loader>
							</Dimmer>
						</Container>
					) : (
						<Container>
							<Container>
								<Header as="h2">Fuel types overview</Header>
								<Grid divided="vertically">
									<Grid.Row columns={dashboardLayout}>
										<Grid.Column>
											<MixChart fuelData={output.generationmix} />
										</Grid.Column>
										<Grid.Column>
											<Divider hidden />
											<Divider hidden />
											<Label>
												<Icon name="time" />
												<Divider hidden fitted />
												Values for the period
												<Divider hidden fitted />
												from :
												<Divider hidden fitted />
												<Label color="yellow">{output.from}</Label>
												<Divider hidden fitted />
												to:
												<Divider hidden fitted />
												<Label color="yellow">{output.from}</Label>
											</Label>
										</Grid.Column>
									</Grid.Row>
								</Grid>
								<Divider />
							</Container>
							<Header as="h2">Fuel types sorted by percentage energy generated</Header>
							<Divider hidden />
							<FuelCard fuelSource={output.generationmix} windowWidth={this.state.width} />
						</Container>
					)}
				</Container>
			);
		};
	};
};

export default Dashboard;
