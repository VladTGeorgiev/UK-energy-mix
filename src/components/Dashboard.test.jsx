import React from 'react';
import Dashboard from './Dashboard.jsx';
import renderer from 'react-test-renderer'

global.fetch = require('jest-fetch-mock');

describe('Dashboard', () => {
	it('should render Loading screen when there is no data loaded', () => {
		const tree = renderer.create(<Dashboard />).toJSON();
		expect(tree).toMatchSnapshot();
	});
});

