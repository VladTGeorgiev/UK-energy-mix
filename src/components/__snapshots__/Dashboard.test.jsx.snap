// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render Loading screen when there is no data loaded 1`] = `
<div
  className="ui center aligned container"
>
  <div
    className="ui hidden divider"
  />
  <h1
    className="ui header"
  >
    UK Energy Mix
  </h1>
  <div
    className="ui divider"
  />
  <div
    className="ui container"
  >
    <div
      className="ui hidden divider"
    />
    <div
      className="ui active transition visible inverted dimmer"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="ui large text loader"
        >
          Loading...
        </div>
      </div>
    </div>
  </div>
</div>
`;
